cpp/src/copying/gather.cu

```diff
@@ -210,121 +213,183 @@ struct column_gatherer {
     // Copy temporary buffers used for in-place gather to destination column
     if (in_place) {
       thrust::copy(rmm::exec_policy(stream)->on(stream), destination_data,
-                   destination_data + num_destination_rows,
-                   static_cast<ColumnType *>(destination_column->data));
+		   destination_data + num_destination_rows,
+		   static_cast<ColumnType *>(destination_column->data));
     }
 
     CHECK_STREAM(stream);
   }
 };
 
-void gather(table const *source_table, gdf_index_type const gather_map[],
-            table *destination_table, bool check_bounds,
-            bool sync_nvstring_category) {
-  CUDF_EXPECTS(nullptr != source_table, "source table is null");
-  CUDF_EXPECTS(nullptr != destination_table, "destination table is null");
-
-  // If the destination is empty, return immediately as there is nothing to
-  // gather
-  if (0 == destination_table->num_rows()) {
-    return;
+template <typename MapType>
+struct index_map : public thrust::unary_function<MapType,MapType>
+{
+  index_map(gdf_size_type source_n_rows) : source_n_rows(source_n_rows){}
+  __device__
+  MapType operator()(MapType in) const
+  {
+    return ((in % source_n_rows) + source_n_rows) % source_n_rows;
```

Comment:
This is just the `pymod` operation from here:

https://github.com/rapidsai/cudf/blob/branch-0.10/cpp/src/binaryop/jit/code/operation.cpp#L162

Unfortunately I couldn't see a way to easily reuse that code.

It's required because Python and C modulo have different behaviour:

https://stackoverflow.com/questions/1907565/c-and-python-different-behaviour-of-the-modulo-operation

Reply:
A comment in the code will suffice.

Reply:
Done.